python/ray/util/client/dataclient.py

```diff
@@ -71,20 +79,36 @@ def _data_main(self) -> None:
                     # This is not being waited for.
                     logger.debug(f"Got unawaited response {response}")
                     continue
+                self.lock.acquire()
                 if response.req_id in self.asyncio_waiting_data:
                     callback = self.asyncio_waiting_data.pop(response.req_id)
+                    self.lock.release()
                     try:
                         callback(response)
                     except Exception:
                         logger.exception("Callback error:")
                 else:
+                    self.lock.release()
                     with self.cv:
                         self.ready_data[response.req_id] = response
                         self.cv.notify_all()
         except grpc.RpcError as e:
             with self.cv:
                 self._in_shutdown = True
+                self._last_exception = e
                 self.cv.notify_all()
+
+            with self.lock:
+                # Abort async requests with the error.
+                err = ConnectionError(
+                    "Failed during this or a previous request. "
+                    f"Exception that broke the connection: {e}")
+                for callback in self.asyncio_waiting_data.values():
+                    callback(err)
```

Comment:
Should we avoid calling this while holding `self.lock`?

Reply:
Good point, refactored.